sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala

```diff
@@ -45,24 +45,39 @@ private[csv] class UnivocityParser(
   // A `ValueConverter` is responsible for converting the given value to a desired type.
   private type ValueConverter = String => Any
 
+  private val shouldHandleCorruptRecord = options.permissive && requiredSchema.exists { f =>
+    f.name == options.columnNameOfCorruptRecord && f.dataType == StringType && f.nullable
+  }
+
+  private val inputSchema = if (shouldHandleCorruptRecord) {
+    StructType(schema.filter(_.name != options.columnNameOfCorruptRecord))
+  } else {
+    schema
+  }
+
   private val valueConverters =
-    schema.map(f => makeConverter(f.name, f.dataType, f.nullable, options)).toArray
+    inputSchema.map(f => makeConverter(f.name, f.dataType, f.nullable, options)).toArray
 
   private val parser = new CsvParser(options.asParserSettings)
 
   private var numMalformedRecords = 0
 
   private val row = new GenericInternalRow(requiredSchema.length)
 
-  private val indexArr: Array[Int] = {
+  private val corruptIndex =
+    requiredSchema.getFieldIndex(options.columnNameOfCorruptRecord).getOrElse(-1)
```

Comment:
So, as to https://github.com/apache/spark/pull/16928/files#r101911027, I suggest to change this to something like ...

```scala
private val corruptFieldIndex = schema.getFieldIndex(options.columnNameOfCorruptRecord)
corruptFieldIndex.foreach { idx =>
  require(schema(idx).dataType == StringType && schema(idx).nullable)
  require(schema(idx).nullable)
  // Corrupted schema should be added at the last.
  require(idx == schema.length - 1)
}
```

Then, it would not require the changes 48L - 57L, 78L - 80L, 191L & 186 and etc.